packages/repo-metadata-lint/src/get-repo-files.ts

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {OctokitType} from './utils/octokit-util';
+import {Storage} from '@google-cloud/storage';
+import {logger} from 'gcf-utils';
+
+/**
+ * This interface is used for comforting typescript's type system to
+ * deal with the response from `octokit.repos.getContent`.
+ */
+interface File {
+  content: string | undefined;
+}
+
+/**
+ * This function is used for comforting typescript's type system to
+ * deal with the response from `octokit.repos.getContent`.
+ */
+function isFile(file: File | unknown): file is File {
+  return (file as File).content !== undefined;
+}
+
+// Load full contents of GitHub repository from commit cache,
+// use this to iterate over individual .repo-metadata.json files.
+// This approach is used for the benefit of mono-repos.
+export class GetRepoFiles {
+  octokit: OctokitType;
+  owner: string;
+  repo: string;
+  bucket: string;
+  storage: Storage;
+  constructor(
+    owner: string,
+    repo: string,
+    bucket: string,
+    octokit: OctokitType
+  ) {
+    this.octokit = octokit;
+    this.owner = owner;
+    this.repo = repo;
+    this.bucket = bucket;
+    this.storage = new Storage();
+  }
+  // Emit any .repo-metadata.json files found in file listing.
+  async *repoMetadata() {
+    const files = await this.getFileListing();
+    if (!files) {
+      logger.warn(`unable to find file listing for ${this.owner}/${this.repo}`);
+      return;
+    }
+    for (const path of files.trim().split(/\r?\n/)) {
+      if (path.toLowerCase().endsWith('.repo-metadata.json')) {
+        const resp = (
+          await this.octokit.rest.repos.getContent({
+            owner: this.owner,
+            repo: this.repo,
+            path,
+          })
+        ).data;
+        if (isFile(resp)) {
+          yield [path, Buffer.from(resp.content, 'base64').toString('utf8')];
+        }
+      }
+    }
+  }
+  // Fetch the file listing from commit cache.
+  private async getFileListing(): Promise<string> {
+    const perPage = 5;
```

Comment:
isn't `5` a kinda silly low number for pagination?

Reply:
We actually want the results from the most recent commit in the history, but if a commit happened within the last 10 minutes or so, it takes a while for DRIFT's pipeline to process it.

The only reason we go back 5 commits is so that we fallback to the previous commit, whicih is likely to have been processed -- could make the number bigger, but I think 5 willi tend to cover us (I landed on this number building the owlbot dashboard).